### PR TITLE
[WIP] fix(addon-editor): support for editing links in the new editor

### PR DESCRIPTION
**Pull request by splincode**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows [Conventional Commits](https://www.conventionalcommits.org/en/)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Refactoring
- [ ] Code style update
- [ ] Build or CI related changes
- [ ] Documentation content changes

## What is the current behavior?

Part of #1302

https://user-images.githubusercontent.com/12021443/161262939-283778f6-239d-4906-8475-c07ef3bdad0a.mov

https://user-images.githubusercontent.com/12021443/161263066-a9a4b139-597b-4541-9b1b-20edb2f7ba5c.mov

https://user-images.githubusercontent.com/12021443/161263227-3a296840-0237-457a-b134-d89511323102.mov

## What is the new behavior?

https://user-images.githubusercontent.com/12021443/161263505-906c5161-4a1f-4d3d-9770-276828c0ed14.mov

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: projects/addon-editor/components/editor-new/editor-new.component.ts [hunk 1]
```diff
@@ -133,11 +150,39 @@ export class TuiEditorNewComponent
         this.editor?.destroy();
     }
 
+    currentFocusedNodeIsAnchor(): boolean {
```
Comment (waterplea) [inline projects/addon-editor/components/editor-new/editor-new.component.ts:153]: Do these methods require being public?

**Review by waterplea — CHANGES_REQUESTED** [review]

File: projects/addon-editor/components/editor-new/editor-new.component.ts (hunk 1 above)
Comment (splincode) [reply projects/addon-editor/components/editor-new/editor-new.component.ts]: fixed

File: projects/addon-editor/components/editor-new/editor-new.component.ts [hunk 2]
```diff
@@ -72,12 +77,20 @@ export class TuiEditorNewComponent
...
+        @Optional()
+        @Inject(TUI_DOCUMENT_OR_SHADOW_ROOT)
+        private readonly shadowRootRef: DocumentOrShadowRoot | null,
     ) {
         super(control, changeDetectorRef);
     }
 
     get dropdownSelectionHandler(): TuiBooleanHandler<Range> {
-        return this.focused ? this.isSelectionLink : ALWAYS_FALSE_HANDLER;
+        return (): boolean =>
```
[the remark this answers is not in the document]
Comment (splincode) [reply projects/addon-editor/components/editor-new/editor-new.component.ts]: fixed

File: projects/addon-editor/components/editor-new/editor-new.component.ts [hunk 3]
```diff
@@ -72,12 +77,20 @@ export class TuiEditorNewComponent
         @Inject(ChangeDetectorRef) changeDetectorRef: ChangeDetectorRef,
         @Inject(TIPTAP_EDITOR) readonly editorLoaded$: Observable<Editor | null>,
         @Inject(TuiTiptapEditorService) readonly editorService: TuiEditor,
+        @Inject(DOCUMENT)
+        private readonly documentRef: Document,
+        @Optional()
+        @Inject(TUI_DOCUMENT_OR_SHADOW_ROOT)
```
[the remark this answers is not in the document]
Comment (splincode) [reply projects/addon-editor/components/editor-new/editor-new.component.ts]: fixed

File: projects/addon-editor/components/editor-new/editor-new.component.ts [hunk 4]
```diff
@@ -72,12 +76,16 @@ export class TuiEditorNewComponent
...
+        @Inject(DOCUMENT)
+        private readonly documentRef: Document,
     ) {
         super(control, changeDetectorRef);
     }
 
     get dropdownSelectionHandler(): TuiBooleanHandler<Range> {
-        return this.focused ? this.isSelectionLink : ALWAYS_FALSE_HANDLER;
+        return this.currentFocusedNodeIsAnchor()
+            ? () => this.isValidSelectionAnchorRange()
```
Comment (waterplea) [inline projects/addon-editor/components/editor-new/editor-new.component.ts:87]: This way you still rebuild a function on each change detection

File: projects/addon-editor/components/editor-new/editor-new.component.ts [hunk 5]
```diff
@@ -137,7 +149,35 @@ export class TuiEditorNewComponent
         return '';
     }
 
-    private readonly isSelectionLink = (): boolean => !!this.editor?.isActive('link');
+    private currentFocusedNodeIsAnchor(): boolean {
+        return !!this.selection?.anchorNode?.parentElement?.closest('a');
```
Comment (waterplea) [inline projects/addon-editor/components/editor-new/editor-new.component.ts:153]: `DropdownSelection` uses `TuiBooleanHandler<Range>` which means the function receives currently selected `Range` as input. You can probably use it to determine all those things and not rely on getting `Selection` manually everytime.

**Review by waterplea — CHANGES_REQUESTED** [review]

File: projects/addon-editor/components/editor-new/editor-new.component.ts (hunk 5 above)
Comment (splincode) [reply projects/addon-editor/components/editor-new/editor-new.component.ts:153]: unfortunately, each time the handler shows only the old value, although the cursor is already in a different place

image
image

Comment (splincode) [reply projects/addon-editor/components/editor-new/editor-new.component.ts:145]: @waterplea can you help me with this issue?)

**Pull request closed without merging**